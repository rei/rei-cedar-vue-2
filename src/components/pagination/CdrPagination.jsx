import clsx from 'clsx';
import IconCaretLeft from '../icon/comps/caret-left';
import IconCaretRight from '../icon/comps/caret-right';
import CdrSelect from '../select/CdrSelect';
import style from './styles/CdrPagination.scss';

export default {
  name: 'CdrPagination',
  components: {
    IconCaretLeft,
    IconCaretRight,
    CdrSelect,
  },
  props: {
    /**
     * Total number of pages. Sometimes the total number of pages is different than total page data
     * and this prop allows for that.
     */
    totalPages: {
      type: Number,
      default: null,
    },
    /**
     * Array of objects. Objects have structure of { page: Number, url: String }
     */
    pages: {
      type: Array,
      required: true,
      validator: (value) => {
        const result = value.every((obj) => {
          if (!Object.prototype.hasOwnProperty.call(obj, 'page')
          || typeof obj.page !== 'number') {
            console.error('Property "page" is missing or is not a number', obj); // eslint-disable-line
            return false;
          } if (!Object.prototype.hasOwnProperty.call(obj, 'url')
          || typeof obj.url !== 'string') {
            console.error('Property "url" is missing or is not a string', obj); // eslint-disable-line
            return false;
          }
          return true;
        });
        return result;
      },
    },
    /** @ignore used for binding v-model, represents the current page */
    value: {
      type: Number,
    },
  },
  data() {
    return {
      currentUrl: '',
      style,
    };
  },
  computed: {
    totalPageData() {
      return this.pages.length;
    },
    prevPageIdx() {
      return this.currentIdx - 1;
    },
    nextPageIdx() {
      return this.currentIdx + 1;
    },
    currentIdx() {
      return this.pages.map(x => x.page).indexOf(this.value);
    },
    /**
     * Creates an array of the pages that should be shown as links with logic for truncation.
     *
     * If total = 20 ([num] indicates current page)
     * [1] 2 3 4 5 ... 20
     * 1 2 3 [4] 5 ... 20
     * 1 ... 4 [5] 6 ... 20
     * 1 ... 15 [16] 17 ... 20
     * 1 ... 16 [17] 18 19 20
     *
     * If total 5 (or anything <= 7)
     * [1] 2 3 4 5
     * 1 2 [3] 4 5
     * 1 2 3 4 [5]
     *
     * Degrades for prev/next only scenario
     * urls = {"4": "url4", "5": "url5", "6": "url6"} (assuming current page is 5)
     * 4 [5] 6
     */
    paginationData() {
      const total = this.totalPageData;
      const current = this.value;
      const delta = 1;
      let range = [];
      let over5 = true;
      let over5remain = true;

      if (total <= 7) {
        // all pages
        return this.pages;
      }

      if (current < 5) {
      // if first 5 pages
        over5 = false;
        // [2-5]
        range = this.pages.slice(1, 5);
      } else if (total - current < 4) {
      // if last 5 pages
        over5remain = false;
        range = this.pages.slice(-5, -1);
      } else {
      // else in between
        for (
          let i = Math.max(2, current - delta);
          i <= Math.min(total - 1, current + delta);
          i += 1
        ) {
          range.push(this.pages[i - 1]);
        }
      }

      if ((current - delta > 2) && over5) {
        range.unshift('&hellip;');
      }
      if ((current + delta < total - 1) && over5remain) {
        range.push('&hellip;');
      }

      range.unshift(this.pages[0]);
      range.push(this.pages[total - 1]);

      return range;
    },
    prevElAttrs() {
      return {
        class: clsx(this.style['cdr-pagination__link'], this.style['cdr-pagination__prev']),
        href: this.pages[this.prevPageIdx].url,
        page: this.pages[this.prevPageIdx].page,
        'aria-label': 'Go to previous Page',
        ref: 'prev-link',
        content: 'Prev',
        iconClass: this.style['cdr-pagination__caret--prev'],
        iconComponent: 'icon-caret-left',
        iconPath: '#caret-left',
      };
    },
    prevEl() {
      return this.value > this.pages[0].page ? (
        <li>
          {this.$scopedSlots.prevLink
            ? this.$scopedSlots.prevLink(this.prevElAttrs)
            : (<a
              class={this.prevElAttrs.class}
              href={this.prevElAttrs.href}
              page={this.prevElAttrs.page}
              aria-label={this.prevElAttrs['aria-label']}
              ref={this.prevElAttrs.ref}
              onClick={e => this.navigate(this.prevElAttrs.page, e)}
            >
              <this.prevElAttrs.iconComponent
                class={this.prevElAttrs.iconClass}
              />
              {this.prevElAttrs.content}
            </a>)
          }
        </li>
      ) : '';
    },
    nextElAttrs() {
      return {
        class: clsx(this.style['cdr-pagination__link'], this.style['cdr-pagination__next']),
        href: this.pages[this.nextPageIdx].url,
        page: this.pages[this.nextPageIdx].page,
        'aria-label': 'Go to next page',
        ref: 'next-link',
        content: 'Next',
        iconClass: this.style['cdr-pagination__caret--next'],
        iconComponent: 'icon-caret-right',
        iconPath: '#caret-right',
      };
    },
    nextEl() {
      return this.value < this.pages[this.totalPageData - 1].page ? (
        <li>
          {this.$scopedSlots.nextLink
            ? this.$scopedSlots.nextLink(this.nextElAttrs)
            : (<a
              class={this.nextElAttrs.class}
              href={this.nextElAttrs.href}
              page={this.nextElAttrs.page}
              aria-label={this.nextElAttrs['aria-label']}
              ref={this.nextElAttrs.ref}
              onClick={e => this.navigate(this.nextElAttrs.page, e)}
            >
              {this.nextElAttrs.content}
              <this.nextElAttrs.iconComponent
                class={this.nextElAttrs.iconClass}
              />
            </a>)
          }

        </li>
      ) : '';
    },
    desktopEl() {
      return this.paginationData.map(n => (
          <li
            key={`${n}-${this.guid()}`}
            class={this.style['cdr-pagination__li--links']}
          >
            {n === '&hellip;'
              && <span
                class={this.style['cdr-pagination__ellipse']}
                domPropsInnerHTML={n}
              />
            }
            {n !== '&hellip;' && this.renderLinkEl(n) }
          </li>
      ));
    },
    mobileEl() {
      return (
        <li class={this.style['cdr-pagination__li--select']}>
          <cdr-select
            vModel={this.currentUrl}
            label="Navigate to page"
            hide-label
            onChange={this.select}
            class={this.style['cdr-pagination__select']}
            ref="select"
          >
            {this.paginationData.map(n => n !== '&hellip;'
              && (<option
                key={`${n}-${this.guid()}`}
                value={n.url}
              >
                Page { n.page }{ this.totalPages === null ? '' : ` of ${this.totalPages}` }
              </option>))}
          </cdr-select>
        </li>
      );
    },
  },
  watch: {
    value() {
      this.currentUrl = this.pages[this.currentIdx].url;
    },
  },
  mounted() {
    this.currentUrl = this.pages[this.currentIdx].url;
  },
  methods: {
    navigate(num, e) {
      this.$emit('change', num, e);
      this.$emit('input', num, e);
    },
    select(url, e) {
      const idx = this.pages.map(x => x.url).indexOf(url);
      const n = this.pages[idx].page;
      this.$emit('select-change', url, e);
      this.navigate(n, e);
    },
    guid() {
      function s4() {
        return Math.floor((1 + Math.random()) * 0x10000)
          .toString(16)
          .substring(1);
      }
      return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
    },
    renderLinkEl(n) {
      const attrs = {
        class: clsx(this.style['cdr-pagination__link'], { current: n.page === this.value }),
        href: n.url,
        'aria-label': n.page === this.value
          ? `Current page, page ${n.page}`
          : `Go to page ${n.page}`,
        'aria-current': n.page === this.value,
        content: n.page,
        page: n.page,
      };

      return (this.$scopedSlots.link ? this.$scopedSlots.link(attrs)
        : <a
          class={attrs.class}
          href={attrs.href}
          aria-label={attrs['aria-label']}
          aria-current={attrs['aria-current']}
          onClick={e => this.navigate(attrs.page, e)}
        >{ attrs.content }</a>
      );
    },
  },
  render() {
    return (
      <nav aria-label="Pagination Navigation">
        <ul class={this.style['cdr-pagination']}>
          {this.prevEl}
          {this.desktopEl}
          {this.mobileEl}
          {this.nextEl}
        </ul>
      </nav>
    );
  },
};
