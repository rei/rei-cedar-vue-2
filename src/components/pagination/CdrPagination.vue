<template>
  <!-- disable lint errors on line length in template -->
  <!-- eslint-disable max-len -->
  <nav aria-label="Pagination Navigation">
    <ul :class="$style['cdr-pagination']">
      <li
        v-if="localCurrent > pages[0].page"
      >
        <a
          :class="[
            $style['cdr-pagination__link'],
            $style['cdr-pagination__prev'],
          ]"
          :href="pages[prevPageIdx].url"
          :aria-label="`Go to previous Page`"
          @click="navigate(pages[prevPageIdx].page, $event)"
          ref="prev-link"
        ><icon-caret-left
          :class="$style['cdr-pagination__caret--prev']"
          size="small"
        />Prev</a>
      </li>
      <!-- Desktop -->
      <li
        v-for="n in paginationData"
        :key="`${n}-${guid()}`"
        :class="$style['cdr-pagination__li--links']"
      >
        <a
          v-if="n !== '&hellip;'"
          :class="[
            $style['cdr-pagination__link'],
            {'current': n.page === localCurrent}
          ]"
          :href="n.url"
          :aria-label=" n.page === localCurrent ? `Current page, page ${n.page}` : `Go to page ${n.page}`"
          :aria-current="n.page === localCurrent"
          @click="navigate(n.page, $event)"
        >{{ n.page }}</a>
        <!-- eslint-disable vue/no-v-html -->
        <span
          v-else
          :class="$style['cdr-pagination__ellipse']"
          v-html="n"
        />
      </li>
      <!-- Mobile -->
      <li :class="$style['cdr-pagination__li--select']">
        <cdr-select
          v-model="currentUrl"
          label="Navigate to page"
          hide-label
          @change="select(currentUrl, ...arguments)"
          :class="$style['cdr-pagination__select']"
          ref="select"
        >
          <!-- eslint-disable vue/no-use-v-if-with-v-for -->
          <option
            v-for="n in paginationData"
            :key="`${n}-${guid()}`"
            :value="n.url"
            v-if="n !== '&hellip;'"
          >Page {{ n.page }}{{ totalPages === null ? `` : ` of ${totalPages}` }}</option>
        </cdr-select>
      </li>
      <li
        v-if="localCurrent < pages[totalPageData - 1].page"
      >
        <a
          :class="[
            $style['cdr-pagination__link'],
            $style['cdr-pagination__next'],
          ]"
          :href="pages[nextPageIdx].url"
          :aria-label="`Go to next page`"
          @click="navigate(pages[nextPageIdx].page, $event)"
          ref="next-link"
        >Next<icon-caret-right
          :class="$style['cdr-pagination__caret--next']"
          size="small"
        />
        </a>
      </li>
    </ul>
  </nav>
</template>

<script>
import { IconCaretLeft, IconCaretRight } from '@rei/cdr-icon';
import { CdrSelect } from '@rei/cdr-select';

export default {
  name: 'CdrPagination',
  components: {
    IconCaretLeft,
    IconCaretRight,
    CdrSelect,
  },
  model: {
    prop: 'currentPage',
    event: 'change',
  },
  props: {
    /**
     * Total number of pages. Sometimes the total number of pages is different than total page data
     * and this prop allows for that.
     */
    totalPages: {
      type: Number,
      default: null,
    },
    /**
     * Array of objects. Objects have structure of { page: Number, url: String }
     */
    pages: {
      type: Array,
      required: true,
      validator: /* istanbul ignore next */ (value) => {
        const result = value.every((obj) => {
          if (!Object.prototype.hasOwnProperty.call(obj, 'page')
          || typeof obj.page !== 'number') {
            console.error('Property "page" is missing or is not a number', obj); // eslint-disable-line
            return false;
          } if (!Object.prototype.hasOwnProperty.call(obj, 'url')
          || typeof obj.url !== 'string') {
            console.error('Property "url" is missing or is not a string', obj); // eslint-disable-line
            return false;
          }
          return true;
        });
        return result;
      },
    },
    /** @ignore used for binding v-model */
    currentPage: {
      type: Number,
    },
  },
  data() {
    return {
      localCurrent: this.currentPage,
      currentUrl: '',
    };
  },
  computed: {
    totalPageData() {
      return this.pages.length;
    },
    prevPageIdx() {
      return this.currentIdx - 1;
    },
    nextPageIdx() {
      return this.currentIdx + 1;
    },
    currentIdx() {
      return this.pages.map(x => x.page).indexOf(this.localCurrent);
    },
    /**
     * Creates an array of the pages that should be shown as links with logic for truncation.
     *
     * If total = 20 ([num] indicates current page)
     * [1] 2 3 4 5 ... 20
     * 1 2 3 [4] 5 ... 20
     * 1 ... 4 [5] 6 ... 20
     * 1 ... 15 [16] 17 ... 20
     * 1 ... 16 [17] 18 19 20
     *
     * If total 5 (or anything <= 7)
     * [1] 2 3 4 5
     * 1 2 [3] 4 5
     * 1 2 3 4 [5]
     *
     * Degrades for prev/next only scenario
     * urls = {"4": "url4", "5": "url5", "6": "url6"} (assuming current page is 5)
     * 4 [5] 6
     */
    paginationData() {
      const total = this.totalPageData;
      const current = this.localCurrent;
      const delta = 1;
      let range = [];
      let over5 = true;
      let over5remain = true;

      if (total <= 7) {
        // all pages
        return this.pages;
      }

      if (current < 5) {
      // if first 5 pages
        over5 = false;
        // [2-5]
        range = this.pages.slice(1, 5);
      } else if (total - current < 4) {
      // if last 5 pages
        over5remain = false;
        range = this.pages.slice(-5, -1);
      } else {
      // else in between
        for (
          let i = Math.max(2, current - delta);
          i <= Math.min(total - 1, current + delta);
          i += 1
        ) {
          range.push(this.pages[i - 1]);
        }
      }

      if ((current - delta > 2) && over5) {
        range.unshift('&hellip;');
      }
      if ((current + delta < total - 1) && over5remain) {
        range.push('&hellip;');
      }

      range.unshift(this.pages[0]);
      range.push(this.pages[total - 1]);

      return range;
    },
  },
  watch: {
    currentPage(v) {
      this.localCurrent = v;
      this.currentUrl = this.pages[this.currentIdx].url;
    },
  },
  mounted() {
    this.currentUrl = this.pages[this.currentIdx].url;
  },
  methods: {
    navigate(num, e) {
      this.$emit('change', num, e);
    },
    select(url, url2, e) {
      const idx = this.pages.map(x => x.url).indexOf(url);
      const n = this.pages[idx].page;
      this.$emit('select-change', url, e);
      this.navigate(n, e);
    },
    guid() {
      function s4() {
        return Math.floor((1 + Math.random()) * 0x10000)
          .toString(16)
          .substring(1);
      }
      return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
    },
  },
};
</script>

<style module>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrPagination.pcss';
</style>

<style>
  @import '@rei/cdr-icon/dist/cdr-icon.css';
</style>
