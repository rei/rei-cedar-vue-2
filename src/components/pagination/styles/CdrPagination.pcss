/* ==========================================================================
  # CdrPagination
  
  All values should map to variables in
  vars/CdrPagination.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrPagination
========================================================================== */

.cdr-pagination {
  display: flex;
  justify-content: center;
  align-items: center;
  list-style: none;
  padding: 0;

  &__li {
    &--links {
      display: none;

      & + & {
        margin-left: $space-1-x;
      }
    }

    &--select {
      display: block;
    }
  }

  &__link {
    @include redwood-display-30();

    color: $text-color-primary-on-light;
    display: block;
    padding: $inset-three-quarter-x-squish;
    text-decoration: none;

    &:global(.current) {
      box-shadow: inset 0 0 0 1px $taken-for-granite;
      border-radius: $radius-softer;

      &:hover,
      &:focus,
      &:active {
        text-decoration: none;
      }
    }

    &:hover,
    &:focus,
    &:active {
      color: $text-color-primary-on-light;
      text-decoration: underline;
    }
  }

  &__ellipse {
    display: block;
    padding: $inset-three-quarter-x-squish;
  }

  &__prev {
    display: flex;
    align-items: center;
    margin-right: $space-1-and-a-half-x;
  }

  &__next {
    display: flex;
    align-items: center;
    margin-left: $space-1-and-a-half-x;
  }

  &__caret {
    &--prev {
      margin-right: $space-eighth-x;
    }

    &--next {
      margin-left: $space-eighth-x;
    }
  }

  &__select {
    max-width: 240px;
  }

  /* Responsive */
  @media (--sm-mq) {
    &__li--links {
      display: block;
    }

    &__li--select {
      display: none;
    }
  }
}
