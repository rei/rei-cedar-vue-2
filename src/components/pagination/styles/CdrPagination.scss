@import '../../../css/settings/index.scss';
@import url('@rei/cedar/dist/style/cdr-icon.css');
@import url('@rei/cedar/dist/style/cdr-select.css');
/* ==========================================================================
  # CdrPagination

  All values should map to variables in
  vars/CdrPagination.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrPagination
========================================================================== */

.cdr-pagination {
  display: flex;
  justify-content: center;
  align-items: center;
  list-style: none;
  padding: 0;

  &__li {
    &--links {
      display: none;

      & + & {
        margin-left: $cdr-space-quarter-x;
      }
    }

    &--select {
      display: block;
    }
  }

  &__link {
    @include cdr-text-utility-sans-300;

    color: $cdr-color-text-primary-lightmode;
    fill: $cdr-color-text-primary-lightmode;
    display: block;
    padding: $cdr-space-inset-three-quarter-x-squish;
    text-decoration: none;
    border-radius: $cdr-radius-softer;

    &:hover,
    &:focus,
    &:active {
      color: $cdr-color-text-primary-lightmode;
      background-color: $cdr-color-background-light;
    }

    &:global(.current) {
      color: $cdr-color-text-link-lightmode;
      position: relative;

      &:after {
        content: '';
        position: absolute;
        bottom: 0;
        left: 50%;
        transform: translateX(-50%);
        height: 2px;
        width: 80%;
        background-color: $cdr-color-text-link-lightmode;
      }

      &:hover,
      &:focus,
      &:active {
        color: $cdr-color-text-link-lightmode;
        text-decoration: none;
      }
    }

    &#{&}--disabled {
      color: $cdr-color-text-disabled-lightmode;
      fill: $cdr-color-text-disabled-lightmode;
      pointer-events: none;

      &:hover,
      &:focus,
      &:active {
        color: $cdr-color-text-disabled-lightmode;
        background-color: transparent;
      }
    }
  }

  &__ellipse {
    display: block;
    padding: $cdr-space-inset-three-quarter-x-squish;
  }

  &__prev {
    display: flex;
    align-items: center;
    margin-right: $cdr-space-one-x;
  }

  &__next {
    display: flex;
    align-items: center;
    margin-left: $cdr-space-one-x;
  }

  &__caret {
    &--prev {
      margin-right: $cdr-space-eighth-x;
    }

    &--next {
      margin-left: $cdr-space-eighth-x;
    }
  }

  /* Responsive */
  @include sm-mq {
    &__li--links {
      display: block;
    }

    &__li--select {
      display: none;
    }
  }
}
