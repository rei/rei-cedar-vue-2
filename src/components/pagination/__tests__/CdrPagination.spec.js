import { mount, shallowMount } from '@vue/test-utils';
import CdrPagination from 'componentdir/pagination/CdrPagination';

function makePages(total, startingAt = 0) {
  const arg = 'page';
  const adjuster = startingAt > 0 ? startingAt : 0;
  const result = [];
  const arr = Array(total).fill().map((_, i) => i + adjuster + 1);
  arr.forEach((n) => {
    const obj = {};
    obj.page = n;
    obj.url = `?${arg}=${n}`;
    result.push(obj);
  });
  return result;
}

function getPageNumArray(input) {
  return input.map((x) => {
    return x === '&hellip;' ? '...' : x.page;
  });
}

function getPrevNextPages(p) {
  if (p === 1 || p === 10) {
    return 2;
  }
  return 3;
}

describe('CdrPagination', () => {
  it('renders correctly', () => {
    const wrapper = mount(CdrPagination, {
      propsData: {
        pages: makePages(20),
        value: 1,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('has a wrapping nav element for a11y', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(20),
        value: 1,
      },
    });
    expect(wrapper.is('nav')).toBe(true);
  });

  it('sorts 20 pages correctly', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(20),
        value: 1,
      },
    });
    let prev = wrapper.find({ref: 'prev-link'});
    let next = wrapper.find({ref: 'next-link'});

    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,2,3,4,5,'...',20]);
    expect(prev.exists()).toBeFalsy();
    expect(next.exists()).toBeTruthy();

    wrapper.setProps({ value: 4 })
    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,2,3,4,5,'...',20]);
    prev = wrapper.find({ ref: 'prev-link' });
    next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeTruthy();

    wrapper.setProps({ value: 20 })
    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,'...',16,17,18,19,20]);
    prev = wrapper.find({ ref: 'prev-link' });
    next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeFalsy();

    wrapper.setProps({ value: 17 })
    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,'...',16,17,18,19,20]);
    prev = wrapper.find({ ref: 'prev-link' });
    next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeTruthy();
  });

  it('sorts 5 pages correctly', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(5),
        value: 1,
      },
    });
    let prev = wrapper.find({ref: 'prev-link'});
    let next = wrapper.find({ref: 'next-link'});
    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,2,3,4,5]);
    expect(prev.exists()).toBeFalsy();
    expect(next.exists()).toBeTruthy();

    wrapper.setProps({ value: 4 })
    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,2,3,4,5]);
    prev = wrapper.find({ ref: 'prev-link' });
    next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeTruthy();

    wrapper.setProps({ value: 5 })
    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1,2,3,4,5]);
    prev = wrapper.find({ ref: 'prev-link' });
    next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeFalsy();
  });

  it('sorts prev/next only pages correctly', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(getPrevNextPages(5), 3),
        value: 5,
      },
    });

    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([4, 5, 6]);
    let prev = wrapper.find({ ref: 'prev-link' });
    let next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeTruthy();
  });

  it('shows next but not prev link when at first page', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(getPrevNextPages(1), -1),
        value: 1,
      },
    });

    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([1, 2]);
    const prev = wrapper.find({ ref: 'prev-link' });
    const next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeFalsy();
    expect(next.exists()).toBeTruthy();
  });

  it('shows prev but not next link when at last page', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(getPrevNextPages(10), 8),
        value: 10,
      },
    });

    expect(getPageNumArray(wrapper.vm.paginationData)).toEqual([9, 10]);
    const prev = wrapper.find({ ref: 'prev-link' });
    const next = wrapper.find({ ref: 'next-link' });
    expect(prev.exists()).toBeTruthy();
    expect(next.exists()).toBeFalsy();
  });

  it('emits events with correct values', () => {
    const wrapper = mount(CdrPagination, {
      propsData: {
        pages: makePages(20),
        value: 5,
      },
    });
    let next = wrapper.find({ ref: 'next-link' });
    let prev = wrapper.find({ ref: 'prev-link' });

    // click next
    next.trigger('click');
    expect(wrapper.emitted().change[0][0]).toBe(6);
    expect(wrapper.emitted().change[0][1] instanceof Event).toBeTruthy();

    // click previous
    prev.trigger('click');
    expect(wrapper.emitted().change[1][0]).toBe(4);
    expect(wrapper.emitted().change[1][1] instanceof Event).toBeTruthy();

    // click a page link
    let link = wrapper.findAll('ul > li > a').at(1);
    link.trigger('click');
    expect(wrapper.emitted().change[2][0]).toBe(1);
    expect(wrapper.emitted().change[2][1] instanceof Event).toBeTruthy();

    // use select
    let options = wrapper.find({ ref: 'select' }).findAll('option')
    options.at(1).setSelected();
    expect(wrapper.emitted()['select-change'][0][0]).toBe('?page=4');
    expect(wrapper.emitted()['select-change'][0][1] instanceof Event).toBeTruthy();
  });

  it('adds "of x" when a total is provided', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(20),
        value: 1,
      },
    });

    let option = wrapper.find({ ref: 'select' }).findAll('option').at(0);
    expect(option.text()).toBe('Page 1');

    wrapper.setProps({ totalPages: 20 });
    option = wrapper.find({ ref: 'select' }).findAll('option').at(0);
    expect(option.text()).toBe('Page 1 of 20');
  });

  it('alows links to be overriden with scoped slots', () => {
    const wrapper = shallowMount(CdrPagination, {
      propsData: {
        pages: makePages(3),
        value: 2,
      },
      scopedSlots: {
        link: '<p slot-scope="link">{{link.href}} {{link.content}}</p>',
        prevLink: '<p slot-scope="link">{{link.href}} {{link.content}}</p>',
        nextLink: '<p slot-scope="link">{{link.href}} {{link.content}}</p>',
      }
    });
    expect(wrapper.text()).toBe('?page=1 Prev?page=1 1?page=2 2?page=3 3Page 1Page 2Page 3?page=3 Next');
  });
});
