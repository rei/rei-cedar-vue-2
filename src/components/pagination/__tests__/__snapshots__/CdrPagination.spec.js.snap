// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrPagination renders correctly 1`] = `
<nav
  aria-label="Pagination Navigation"
>
  <ul
    class="cdr-pagination"
  >
    <li
      class="cdr-pagination__li--links"
    >
      <a
        aria-current="true"
        aria-label="Current page, page 1"
        class="cdr-pagination__link current"
        href="?page=1"
      >
        1
      </a>
    </li>
    <li
      class="cdr-pagination__li--links"
    >
      <a
        aria-label="Go to page 2"
        class="cdr-pagination__link"
        href="?page=2"
      >
        2
      </a>
    </li>
    <li
      class="cdr-pagination__li--links"
    >
      <a
        aria-label="Go to page 3"
        class="cdr-pagination__link"
        href="?page=3"
      >
        3
      </a>
    </li>
    <li
      class="cdr-pagination__li--links"
    >
      <a
        aria-label="Go to page 4"
        class="cdr-pagination__link"
        href="?page=4"
      >
        4
      </a>
    </li>
    <li
      class="cdr-pagination__li--links"
    >
      <a
        aria-label="Go to page 5"
        class="cdr-pagination__link"
        href="?page=5"
      >
        5
      </a>
    </li>
    <li
      class="cdr-pagination__li--links"
    >
      <span
        class="cdr-pagination__ellipse"
      >
        â€¦
      </span>
    </li>
    <li
      class="cdr-pagination__li--links"
    >
      <a
        aria-label="Go to page 20"
        class="cdr-pagination__link"
        href="?page=20"
      >
        20
      </a>
    </li>
    <li
      class="cdr-pagination__li--select"
    >
      <div
        class="cdr-pagination__select"
      >
        <div
          class="cdr-select-wrap"
        >
          <select
            aria-label="Navigate to page"
            class="cdr-select  cdr-select cdr-select__prompt"
            id="2"
          >
            <option
              value="?page=1"
            >
              Page 1
            </option>
            <option
              value="?page=2"
            >
              Page 2
            </option>
            <option
              value="?page=3"
            >
              Page 3
            </option>
            <option
              value="?page=4"
            >
              Page 4
            </option>
            <option
              value="?page=5"
            >
              Page 5
            </option>
            <option
              value="?page=20"
            >
              Page 20
            </option>
          </select>
          <svg
            class="cdr-icon  cdr-select__caret"
            role="presentation"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 0 0-1.414-1.414L12 13.586 7.705 9.29a1 1 0 0 0-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
            />
          </svg>
        </div>
      </div>
    </li>
    <li>
      <a
        aria-label="Go to next page"
        class="cdr-pagination__link cdr-pagination__next"
        href="?page=2"
        page="2"
      >
        Next
        <svg
          class="cdr-icon  cdr-pagination__caret--next"
          role="presentation"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 12a.997.997 0 0 0-.288-.702l-5.005-5.005a1 1 0 0 0-1.414 1.414L13.585 12 9.29 16.295a1 1 0 0 0 1.417 1.412l4.98-4.98A.997.997 0 0 0 16 12z"
          />
        </svg>
      </a>
    </li>
  </ul>
</nav>
`;
